src/clients/c++/experimental_api_v2/library/common.h

```diff
@@ -0,0 +1,561 @@
+  Error GetNext(const uint8_t** buf, size_t* input_bytes, bool* end_of_input);
+
+ private:
+  InferInput(
+      const std::string& name, const std::vector<int64_t>& dims,
+      const std::string& datatype);
+
+  std::string name_;
+  std::vector<int64_t> shape_;
+  std::string datatype_;
+  size_t byte_size_;
+  size_t total_send_byte_size_;
+
+  size_t bufs_idx_, buf_pos_;
+  std::vector<const uint8_t*> bufs_;
+  std::vector<size_t> buf_byte_sizes_;
+
+  // Used only for STRING type tensors set with SetFromString(). Hold
+  // the "raw" serialization of the string values for each index
+  // that are then referenced by 'bufs_'. A std::list is used to avoid
+  // reallocs that could invalidate the pointer references into the
+  // std::string objects.
+  std::list<std::string> str_bufs_;
+
+  // Used only if working with Shared Memory
+  enum IOType { NONE, RAW, SHARED_MEMORY };
+  IOType io_type_;
+  std::string shm_name_;
+  size_t shm_offset_;
+};
+
+//==============================================================================
+/// An InferRequestedOutput object is used to describe the requested model
+/// output for inference.
+///
+class InferRequestedOutput {
+ public:
+  /// Create a InferRequestedOutput instance that describes a model output being
+  /// requested.
+  /// \param infer_output Returns a new InferOutputGrpc object.
+  /// \param name The name of output being requested.
+  /// \param class_count The number of classifications to be requested. The
+  /// default value is 0 which means the classification results are not
+  /// requested.
+  /// \return Error object indicating success or failure.
+  static Error Create(
+      InferRequestedOutput** infer_output, const std::string& name,
+      const size_t class_count = 0);
+
+  /// Get the name of output associated with this object
+  /// \param name Returns the name of output.
+  /// \return Error object indicating success or failure of the
+  /// request.
+  const std::string& Name() const { return name_; }
+
+  /// Get the number of classifications requested.
+  /// \return The number of classification.
+  size_t ClassCount() const { return class_count_; }
+
+  /// Returns whether the output is requested in shared memory or
+  //  not.
+  /// \return The truth value of wether input is set with shared
+  /// memory.
+  bool IsSharedMemory() const { return (io_type_ == SHARED_MEMORY); }
+
+
+  /// Set the output tensor data to be written to specified shared
+  /// memory region.
+  /// \param region_name The name of the shared memory region.
+  /// \param byte_size The size of data in bytes.
+  /// \param offset The offset in shared memory region. Default value is 0.
+  /// \return Error object indicating success or failure of the
+  /// request.
+  Error SetSharedMemory(
+      const std::string& region_name, const size_t byte_size,
+      const size_t offset = 0);
+
+  // Copy the shared memory key, offset and batch_byte_size
+  Error SharedMemoryInfo(
```

Comment:
Is this the inverse of SetSharedMemory? Why do we need to provide it? In general think about making the API as minimal as possible.